File: typescript/src/queue/queue.ts
// vim: noexpandtab

export class EmptyQueue extends Error {
	constructor(message?: string) {
		super(message);
	}
}

export class QNode<T> {
	public value: T;
	public next?: QNode<T>;

	constructor(item: T) {
		this.value = item;
	}
}

export class Queue<T> {
	public length: number;
	private head?: QNode<T>;
	private tail?: QNode<T>;

	constructor() {
		this.length = 0;
		this.head = undefined;
		this.tail = undefined;
	}

	/**
	 * Appends item to end of queue
	 */
	enqueue(item: T) {
		const node = new QNode(item);

		this.length++;

		if (!this.tail) {
			this.tail = node;
			this.head = node;
			return;
		}

		this.tail.next = node;
		this.tail = node;
	}

	/**
	 * Removes and returns first item of queue or an error
	 * @throws {EmptyQueue} - If/when queue is empty
	 */
	deque(): T {
		if (!this.head) {
			throw new EmptyQueue('Queue is empty');
		}

		this.length--;

		const head = this.head;
		this.head = this.head.next;

		// free
		head.next = undefined;

		return head.value;
	}

	/**
	 * Returns first value of queue without mutation
	 * @throws {EmptyQueue} - If/when queue is empty
	 */
	peek(): T | undefined {
		if (this.head) {
			return this.head.value;
		}
		throw new EmptyQueue('Queue is empty');
	}
}

File: go/queue/queue.go
package queue

import "errors"

type Node[T any] struct {
	value T
	next  *Node[T]
}

type Queue[T any] struct {
	Length uint
	head   *Node[T]
	tail   *Node[T]
}

/**
 * Appends item to end of queue
 */
func (queue *Queue[T]) Enqueue(item T) {
	node := Node[T]{
		value: item,
	}

	queue.Length++

	if queue.Length == 1 {
		queue.tail = &node
		queue.head = &node
		return
	}

	queue.tail.next = &node
	queue.tail = &node
}

/**
 * Removes and returns first item of queue or an error
 */
func (queue *Queue[T]) Deque() (T, error) {
	if queue.Length == 0 {
		var result T
		return result, errors.New("Queue is empty")
	}

	queue.Length--
	head := queue.head
	queue.head = queue.head.next

	return head.value, nil
}

/**
 * Returns first value of queue without mutation
 */
func (queue *Queue[T]) Peek() T {
	return queue.head.value
}
